### PR TITLE
Added support for async initialization

### PR DESCRIPTION
**Pull request by helto4real**
## Proposed change

Adds the possibility for applications to init async operations.
- It is InitializeAsync is called on loading of the applications and will be awaited
- A warning if init takes too long
- A cancellationtoken will be provided to use when application is unloaded

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code compiles without warnings (code quality chek)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/net-daemon/docs

### CODE REVIEW

File: src/AppModel/NetDaemon.AppModel.Tests/Fixtures/Local/LocalAppWithInitializeAsync.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+namespace LocalApps;
+
+[NetDaemonApp]
+public class MyAppLocalAppWithInitializeAsync : IInitializableAsync
```
Comment (FrankBakkerNl) [inline src/AppModel/NetDaemon.AppModel.Tests/Fixtures/Local/LocalAppWithInitializeAsync.cs:4]: IAsyncInitializable is more consistent with IAsyncDisposable and with the current version

File: src/AppModel/NetDaemon.AppModel/Internal/Context/ApplicationContext.cs [hunk 2]
```diff
@@ -2,6 +2,7 @@ namespace NetDaemon.AppModel.Internal;
 
 internal sealed class ApplicationContext
 {
+    private readonly CancellationTokenSource _cancelTokenSource = new();
```
Comment (FrankBakkerNl) [inline src/AppModel/NetDaemon.AppModel/Internal/Context/ApplicationContext.cs:5]: This seems to never be fired, maybe better to pass a token into InitializeAsync()

Comment (helto4real) [reply src/AppModel/NetDaemon.AppModel/Internal/Context/ApplicationContext.cs:5]: Yea I forgot to fire it on DisposeAsync. Good catch

File: src/AppModel/NetDaemon.AppModel.Tests/Fixtures/Local/LocalAppWithInitializeAsync.cs (hunk 1 above)
Comment (helto4real) [reply src/AppModel/NetDaemon.AppModel.Tests/Fixtures/Local/LocalAppWithInitializeAsync.cs]: Lool correct!

File: src/AppModel/NetDaemon.AppModel/Internal/Context/ApplicationContext.cs [hunk 3]
```diff
@@ -23,18 +24,29 @@ public ApplicationContext(Type appType, IServiceProvider serviceProvider)
...
+            _cancelTokenSource.Cancel();
+
         if (Instance is IAsyncDisposable asyncDisposable) await asyncDisposable.DisposeAsync().ConfigureAwait(false);
 
         else if (Instance is IDisposable disposable) disposable.Dispose();
 
         if (_serviceScope is IAsyncDisposable serviceScopeAsyncDisposable)
             await serviceScopeAsyncDisposable.DisposeAsync().ConfigureAwait(false);
+
+        _cancelTokenSource.Dispose();
```
Comment (FrankBakkerNl) [inline src/AppModel/NetDaemon.AppModel/Internal/Context/ApplicationContext.cs:50]: Maybe do this first so the init is cancelled before we dispose

**Review by FrankBakkerNl — CHANGES_REQUESTED** [review]

File: src/AppModel/NetDaemon.AppModel/Internal/Context/ApplicationContext.cs (hunk 3 above)
Comment (helto4real) [reply src/AppModel/NetDaemon.AppModel/Internal/Context/ApplicationContext.cs:50]: Yea that is rifht. The cancellation token could be used some other places too in long running processes.

Comment (helto4real) [reply src/AppModel/NetDaemon.AppModel/Internal/Context/ApplicationContext.cs:50]: I all ready do Cancel before. I do not want to dispose it so code use disposed cancellation token. First I cancel, do asyncdispose call then dispose the  cancellation source.  So I think that should be right?

**Review by FrankBakkerNl — APPROVED** [review]

**Pull request merged**